internal/catalogmetadata/cache/cache.go

```diff
@@ -0,0 +1,143 @@
+	return &filesystemCache{
+		cachePath:              cachePath,
+		mutex:                  sync.RWMutex{},
+		client:                 client,
+		cacheDataByCatalogName: map[string]cacheData{},
+	}
+}
+
+// cacheData holds information about a catalog
+// other than it's contents that is used for
+// making decisions on when to attempt to refresh
+// the cache.
+type cacheData struct {
+	ResolvedRef string
+}
+
+// FilesystemCache is a cache that
+// uses the local filesystem for caching
+// catalog contents. It will fetch catalog
+// contents if the catalog does not already
+// exist in the cache.
+type filesystemCache struct {
+	mutex                  sync.RWMutex
+	cachePath              string
+	client                 *http.Client
+	cacheDataByCatalogName map[string]cacheData
+}
+
+// FetchCatalogContents implements the client.Fetcher interface and
+// will fetch the contents for the provided Catalog from the filesystem.
+// If the provided Catalog has not yet been cached, it will make a GET
+// request to the Catalogd HTTP server to get the Catalog contents and cache
+// them. The cache will be updated automatically if a Catalog is noticed to
+// have a different resolved image reference.
+// The Catalog provided to this function is expected to:
+// - Be non-nil
+// - Have a non-nil Catalog.Status.ResolvedSource.Image
+// This ensures that we are only attempting to fetch catalog contents for Catalog
+// resources that have been successfully reconciled, unpacked, and are being served.
+// These requirements help ensure that we can rely on status conditions to determine
+// when to issue a request to update the cached Catalog contents.
+func (fsc *filesystemCache) FetchCatalogContents(ctx context.Context, catalog *catalogd.Catalog) (io.ReadCloser, error) {
+	if catalog == nil {
+		return nil, fmt.Errorf("error: provided catalog must be non-nil")
+	}
+
+	if catalog.Status.ResolvedSource == nil {
+		return nil, fmt.Errorf("error: catalog %q has a nil status.resolvedSource value", catalog.Name)
+	}
+
+	if catalog.Status.ResolvedSource.Image == nil {
+		return nil, fmt.Errorf("error: catalog %q has a nil status.resolvedSource.image value", catalog.Name)
+	}
+
+	cacheDir := filepath.Join(fsc.cachePath, catalog.Name)
+	cacheFilePath := filepath.Join(cacheDir, "data.json")
+
+	fsc.mutex.RLock()
+	if data, ok := fsc.cacheDataByCatalogName[catalog.Name]; ok {
+		if catalog.Status.ResolvedSource.Image.Ref == data.ResolvedRef {
+			fsc.mutex.RUnlock()
+			return os.Open(cacheFilePath)
+		}
+	}
+	fsc.mutex.RUnlock()
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, catalog.Status.ContentURL, nil)
+	if err != nil {
+		return nil, fmt.Errorf("error forming request: %s", err)
+	}
+
+	resp, err := fsc.client.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("error performing request: %s", err)
+	}
+	defer resp.Body.Close()
+
+	switch resp.StatusCode {
+	case http.StatusOK:
```

Comment:
I suspect the `switch` was here for `If-Modified-Since` implementation where we needed to handle `304` as well.

What do you think about to geting rid of it to avoid nesting?

E.g.

```go
if resp.StatusCode != http.StatusOK {
	return nil, fmt.Errorf("received unexpected response status code %d", resp.StatusCode)
}

fsc.mutex.Lock()
defer fsc.mutex.Unlock()
// ...
```

Reply:
Good point, will update